[Quince Backport] feat: link LTI Provider launches to authenticated users

tecoholic opened the pull request:
<!--

🌳🌳
🌳🌳🌳🌳         🌳 Note: Quince is in support. Fixes you make on master may still be needed on Quince.
    🌳🌳🌳🌳     If so, make another pull request against the open-release/quince.master branch,
🌳🌳🌳🌳         or ask in the #wg-build-test-release Slack channel if you have any questions or need help.
🌳🌳

🌴🌴🌴🌴🌴🌴     🌴 Note: the Palm release is still supported.
                Please consider whether your change should be applied to Palm as well.

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

Quince backport of https://github.com/openedx/edx-platform/pull/33310


Useful information to include:
- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
"Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these
changes.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.
- If your [database migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) can't be rolled back easily.

openedx-webhooks commented:
Thanks for the pull request, @tecoholic! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

tecoholic committed:
feat: link LTI Provider launches to authenticated users

With this change, the platform users who access content via LTI will be
automatically linked to their platform account instead of the new (anonymous)
one. The following conditions need to be met:
* The `LtiConsumer` should be configured to auto-link the users via email.
* The LTI Consumer should share the user's email using the
`lis_person_contact_email_primary` parameter in the LTI Launch POST data.

This also replaces the one-to-one relationship of the `User` and `LtiUser`
with one-to-many. This way, multiple `LtiUser` objects can refer to the same
`edx_user`. With the auto-linking, multiple LTI Consumers can create
independent `LtiUser` objects with the same `edx_user`.

Co-authored-by: Piotr Surowiec <piotr@surowiec.it>
(cherry picked from commit 5b2f012accd13ce89791d0ca3088b9f159c9f926)

Agrendalath reviewed (approved):
:+1:

* [x] I tested this: checked that the diff matches https://github.com/openedx/edx-platform/pull/33310
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here [https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5541071501.log](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5541071501.log)

Please check the settings and requirements and retry deployment by updating the pull request or posting a `/grove sandbox update` comment.

cmltaWt0 commented:
@tecoholic I've read through the original PR. It's an awesome work!

There is a great [testing instruction](https://github.com/openedx/edx-platform/pull/33310#issue-1906863227) in the original PR.
I think we have to merge it right now.

@mariajgrimaldi are you agree?

tecoholic commented:
@cmltaWt0 Thank you. @mariajgrimaldi Kindly let me know if you need anything while testing.

openedx-webhooks commented:
@tecoholic 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/33656&entry.1484655318=https://github.com/tecoholic&entry.752974735=2023-11-06+14:35&entry.1917517419=2023-11-21+18:39&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

cmltaWt0 commented:
**Tested:**
- local images build
- basic LTI Provider functionality

**Not tested:**
- real Canvas integration

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.
